Chinese translation  domain driven design implementation guide (#6505)
abpframework/abp: docs/zh-Hans/Domain-Driven-Design-Implementation-Guide.md

* `Application` 依赖`Application.Contracts`项目，因为此项目需要实现应用服务的接口及接口使用的DTOs。另外也依赖`Domain`项目，因为应用服务的实现必须依赖领域层中的对象。
* `EntityFrameworkCore` 依赖`Domain`项目，因为此项目需要将领域对象（实体或值对象）映射到数据库的表，另外还需要实现`Domain`项目中的仓储接口。
* `HttpApi` 依赖`Application.Contacts`项目，因为Controllers需要注入应用服务。
* `HttpApi.Client` 依赖`Application.Contacts`项目，因为此项目需要是使用应用服务。
* `Web` 依赖`HttpApi`项目，因为此项目对外提供HTTP APIs。另外Pages或Components 需要使用应用服务，所以还间接依赖了`Application.Contacts`项目

#### 虚线依赖

您在上图中会发现用虚线表示了另外两个依赖。`Web`项目依赖了 `Application` and `EntityFrameworkCore`，理论上`Web`不应该依赖这两个项目，但实际上依赖了。原因如下：

`Web`是最终的运行程序，是负责托管Web的宿主，它在运行时需要**应用服务和仓储的真正实现类**。

这种依赖关系的设计，可能会让您有机会在展现层直接使用到EF Core的对象，**应该严格禁止这样的做法**。如果想在解决方案分层上规避这种问题，有下面两种方式，相对复杂一些：

* 将`Web`项目类型改为razor类库，并创建一个新项目，比如`Web.Host`，`Web.Host`依赖`Web`、`Application`、`EntityFrameworkCore`三个项目，并作为Web宿主程序运行。注意，不要写任何与UI相关的代码，只是作为**宿主运行**。
* 在`Web`项目中移除对`Application`和`EntityFrameworkCore`的引用，`Web`在启动时，再动态加载程序集``IssueTracking.Application.dll`和`IssueTracking.EntityFrameworkCore.dll`。可以使用ABP框架的[插件模块](PlugIn-Modules.md)来动态加载程序集。

### DDD模式的应用程序执行顺序

下图展示了基于DDD模式下的Web应用程序执行顺序：

![](images/domain-driven-design-web-request-flow.png)

* 通常由UI（用例）发起一个HTTP请求到服务器。

> * 由展现层（或分布式服务层）中的一个*MVC Controller*或*Razor Page Handler*处理请求，在这次阶段可以执行一些AOP逻辑（授权，验证，异常处理等），*MVC Controller*或*Razor Page Handler*调用注入的应用服务接口，并返回其调用后的结果（DTOs）。.

Reviewer: ```suggestion
* 由展现层（或分布式服务层）中的一个*MVC Controller*或*Razor Page Handler*处理请求，在这个阶段可以执行一些AOP逻辑（授权，验证，异常处理等），*MVC Controller*或*Razor Page Handler*调用注入的应用服务接口，并返回其调用后的结果（DTOs）。.
```

领域层和应用层应该与*Database Provider / ORM*无关。领域层和应用层仅依赖仓储接口，并且仓储接口不依赖特定的ORM对象。

原因如下：

1. 未来领域层或应用层的基础设施会发生改变，例如，需要支持另外一种数据库类型，因此需要保持**领域层或应用层的基础设施是独立的**。
2. 将基础设施的实现隐藏在仓储中，使得领域层或应用层更**专注于业务逻辑代码**。
3. 可以通过模拟仓储接口，使得自动化测试更为方便。

> 关于此原则， `EntityFrameworkCore`项目只被启动程序项目所引用，解决方案中其它项目均未引用。

##### 关于数据库独立原则的讨论

**原因1**会非常影响您**领域对象的建模**（特别是实体间的关系）及**应用程序的代码**。假如，开始选择了关系型数据库，并使用了[Entity Framework Core](Entity-Framework-Core.md)，后面尝试切换到[MongoDB](MongoDB.md)，那么 **EF Core 中一些非常用的特性**您就不能使用了，例如：

* 无法使用[变更追踪](https://docs.microsoft.com/en-us/ef/core/querying/tracking) ，因为*MongoDB provider*没有提供此功能，因此，您始终需要显式的更新已变更的实体。
* 无法在不同的聚合间使用[导航属性](https://docs.microsoft.com/en-us/ef/core/modeling/relationships)，因为文档型数据库是不支持的。有关更多信息，请参见“规则：聚合间仅通过Id关联”。

如果您认为这些功能对您很**重要**，并且您永远都不会**离开** *EF Core*，那么我们认为你可以忽略这一原则。假如您在设计实体关系时使用了*EF Core*，您甚至可以在应用层引用*EF Core Nuget*包，并直接使用异步的LINQ扩展方法，例如 `ToListAsync()`（有关更多信息，请参见[仓储](Repositories.md)文档中的*IQueryable*和*Async Operations*）。

但是我们仍然建议采用仓储模式来隐藏基础设施中实现过程。

#### 展现层技术无关原则

展现层技术（UI框架）时现代应用程序中最多变的部分之一。**领域层和应用层**应该对展现层所采用的技术或框架**一无所知**。使用ABP启动模板就非常容易实现此原则。

在某些情况下，您可能需要在应用层和展现层中写重复的逻辑，例如，参数验证和授权检查。展现层检查出于**用户体验**，应用层或领域层检查出于**数据安全性**和**数据完整性**。

> #### 关注状态的变化而报表/查询(个人觉得更适合)

Reviewer: 原文`Focus on the State Changes, Not Reporting`愿意应该是`关注状态变化, 而不是XXX`. 有没有更好的翻译？

        public IssueAppService(
            IssueManager issueManager,
            IRepository<Issue, Guid> issueRepository,
            IRepository<AppUser, Guid> userRepository)
        {
            _issueManager = issueManager;
            _issueRepository = issueRepository;
            _userRepository = userRepository;
        }

        [Authorize]
        public async Task AssignAsync(IssueAssignDto input)
        {
            var issue = await _issueRepository.GetAsync(input.IssueId);
            var user = await _userRepository.GetAsync(input.UserId);

            await _issueManager.AssignToAsync(issue, user);

            await _issueRepository.UpdateAsync(issue);
        }
    }
}
````

应用服务的方法通常包含三个步骤：

1. 从数据库获取用例所需的领域对象。
2. 使用领域对象（领域服务，实体等）执行业务逻辑。
3. 将实体的变更持久化至数据库。

> 如果使用的是EF Core，第三步不是必须的，因为EF Core有追踪实体变化的功能。如果要利用此功能，请参阅上面的”关于数据库独立原则的讨论“章节。

`IssueAssignDto` 是本示例中一个简单的DTO对象：

````csharp
using System;

namespace IssueTracking.Issues
{
    public class IssueAssignDto
    {
        public Guid IssueId { get; set; }
        public Guid UserId { get; set; }
    }
}
````

### 数据传输对象

[DTO](Data-Transfer-Objects.md)是应用层与展现层间传输数据的简单对象。应用服务方法获取并返回Dto。

#### DTO通用原则和最佳实践

> * DTO应该是**可被序列化**的。因为大所属情况下，DTO是通过网络传输的，因此它应该具有**无参的构造函数**。

Reviewer: ```suggestion
* DTO应该是**可被序列化**的。因为大多数情况下，DTO是通过网络传输的，因此它应该具有**无参的构造函数**。
```

* 不应该包含任何**业务逻辑**。
* **切勿**继承或引用**实体**。

> **输入DTOs**（应用服务方法的参数）与 **输出DTO** （应用服务方法的返回对象）具有不同的作用，因此，它们应该区别对待。 

Reviewer: ```suggestion
**输入DTOs**（应用服务方法的参数）与 **输出DTOs** （应用服务方法的返回对象）具有不同的作用，因此，它们应该区别对待。 
```

#### 输入DTOs 最佳实践

##### 不要在输入DTOs中定义不使用的属性

**仅**在输入DTOs中定义用例**所需要的属性**！否则，会造成调用应用服务的客户端产生困惑。

这个规则好像没什么必要，因为没人会在方法参数（输入DTOs）中添加无用的属性。但是，有时候，特别是在重用DTO时，输入DTOs会包含无用的属性。

##### 不要重用输入DTOs

**为每个用例**（应用服务的方法）单独定义一个**专属的输入DTO**。否则，在一些情况下，会添加一些不被使用的属性，这样就违反上面的规则：不要在输入DTOs中定义不使用的属性。

> 在两个用例中重用相同的DTO似乎很有吸引力，因为它们的属性是一模一样的。现阶段它们是一样的，但是随着业务变化，可能它们会产生差异，届时您可能还是需要进行拆分。**代码的重复相对用例间的耦合，可能是更好的做法**。

Reviewer: ```suggestion
在两个用例中重用相同的DTO似乎很有吸引力，因为它们的属性是一模一样的。现阶段它们是一样的，但是随着业务变化，可能它们会产生差异，届时您可能还是需要进行拆分。**和用例间的耦合相比，代码的复制可能是更好的做法**。
```

{
    public Guid Id { get; set; }
    public string Password { get; set; }
}
````

虽然编写了更多的代码，但是这样可维护性更高。

**例外情况：**该规则有一些例外的情况，例如，您想开发两个方法，它们共用相同的输入DTO（通过继承或重用），有一个报表页面有多个过滤条件，多个应用服务使用相同的输入参数返回不同的结果（如，大屏展示数据，Excel报表，csv报表）。这种情况下，您是需要修改一个参数，多个应用服务都应该一起被修改。

##### 输入DTO中验证逻辑

- 仅在DTO内执行**简单验证**。使用数据注解验证属性或通过`IValidatableObject` 方式。
- **不要执行领域验证**。例如，不要在DTO中检查用户名是否唯一的验证。

**示例：使用注解方式**

````csharp
using System.ComponentModel.DataAnnotations;

namespace IssueTracking.Users
{
    public class UserCreationDto
    {
        [Required]
        [StringLength(UserConsts.MaxUserNameLength)]
        public string UserName { get; set; }

        [Required]
        [EmailAddress]
        [StringLength(UserConsts.MaxEmailLength)]
        public string Email { get; set; }

        [Required]
        [StringLength(UserConsts.MaxEmailLength,
        MinimumLength = UserConsts.MinPasswordLength)]
        public string Password { get; set; }
    }
}
````

当输入无效时，ABP框架会自动验证输入DTO，抛出`AbpValidationException`异常，并向客户返回`400`的HTTP状态码。

> 一些开发人员认为最好将验证规则和DTO分离。我们认为声明性（数据注解）方式是比较实用的，不会引起任何设计问题。如果您喜欢其它方式，ABP还支持[FluentValidation继承](FluentValidation.md)。有关所有验证的详细文档，请参见[验证文档](Validation.md)。

#### 输出DTO最佳实践

> * 保持输出DTO的**个数较少**，尽可能**重用输入DTO**（例外：不要将输入DTO作为输出DTO）。

Reviewer: ```suggestion
* 保持**数量较少**的输出DTO，尽可能**重用输出DTO**（例外：不要将输入DTO作为输出DTO）。
```

> * 输入DTO可以包含比**用例需要的属性更多的属性**。

Reviewer: ```suggestion
* 输出DTO可以包含比用例需要的属性**更多**的属性。
```

}
````

> 这里我们没有使用异步方式，是为了示例更清晰，您实际代码中应该使用异步方式）

上面的示例代码中，每个方法都返回了不同的DTO类型，这样处理，会导致查询数据，映射对象都会有很多重复的代码。

应用服务`IUserAppService` 可以简化成如下代码：

````csharp
public interface IUserAppService : IApplicationService
{
    UserDto Get(Guid id);
    List<UserDto> GetList();
    UserDto Create(UserCreationDto input);
    UserDto Update(UserUpdateDto input);
}
````

只需使用一个DTO对象

````csharp
public class UserDto
{
    public Guid Id { get; set; }
    public string UserName { get; set; }
    public string Email { get; set; }
    public DateTime CreationTime { get; set; }
    public List<string> Roles { get; set; }
}
````

* 删除 `GetUserNameAndEmail`和`GetRoles` 方法，因为，返回的DTO中已经包含了对应的信息。
* `GetList`方法的返回的泛型类型与`Get`方法的返回类型一致。
* `Create` 与 `Update`的返回类型都是 `UserDto`。

如上所述，使用相同的DTO有很多优点。例如，我们在UI上使用**表格**展现用户集合，再用户数据更新后，我们可以获取到返回对象，并对**表格数据源进行更新**。因此，我们无需再次调用`GetList`来获取全部数据。这就是我们为什么建议`Create` 与 `Update`方法都返回相同`UserDto`的原因。

##### 讨论

输出DTO的建议并不适用于所有情况。出于**性能**原因，我们可以忽略这些建议，尤其是在返回**大量数据**、为UI定制、**并发量较高**时。

在这些情况下，您可以定制仅包含**必要信息的DTO**。上面的建议只适用于额外多些属性并**不会损失太多性能**，并关注代码**可维护**的应用系统。

#### 对象映射到对象

当两个对象具有相同或相似的属性，自动将[对象映射到对象](Object-To-Object-Mapping.md)是一种将值从一个对象复制到另外一个对象非常有用的方法。

> DTO和实体通常具有相同或相似的属性，您经常需要从一个实体创建一个DTO对象。于手动映射相比，基于[AutoMapper](http://automapper.org/)的ABP[对象映射系统](Object-To-Object-Mapping.md)，更加方便简单。

Reviewer: ```suggestion
DTO和实体通常具有相同或相似的属性，您经常需要从一个实体创建一个DTO对象。相较于手动映射，基于[AutoMapper](http://automapper.org/)的ABP[对象映射系统](Object-To-Object-Mapping.md)，更加方便简单。
```

> * **仅**在**实体=>输出DTO**的时候使用自动映射。**Use** auto object mapping only for **Entity to output DTO** mappings.

Reviewer: ```suggestion
* **仅**在**实体=>输出DTO**的时候使用自动映射。
```

> * 不要在**输入DTO=>Entity**的时候自动自动映射。

Reviewer: ```suggestion
* 不要在**输入DTO=>Entity**的时候使用自动映射。
```

        if (input.AssignedUserId.HasValue)
        {
            var user = await _userRepository.GetAsync(input.AssignedUserId.Value);
            await _issueManager.AssignToAsync(issue, user);
        }

        // Save
        await _issueRepository.InsertAsync(issue);

        // Return a DTO represents the new Issue
        return ObjectMapper.Map<Issue, IssueDto>(issue);
    }    
}

// *** IssueCreationDto class ***
public class IssueCreationDto
{
    public Guid RepositoryId { get; set; }
    [Required]
    public string Title { get; set; }
    public Guid? AssignedUserId { get; set; }
    public string Text { get; set; }
}
````

##### 讨论：为什么`IssueManager`中没有执行`Issue`的保存？

您可能会问“**为什么`IssueManager`中没有执行`Issue`的保存？**”。我们认为这是应用服务的职责。

因为，应用服务可能在保存`Issue`对象之前，需要对其它对象进行修改。如果领域服务执行了保存，那么*保存*操作就是重复的。

* 会触发两次数据库会交互，这会导致性能损失。
* 需要额外添加显式的事务来包含这两个操作，才能保证数据一致性。
* 如果因为业务规则取消了实体的创建，则应该在数据库事务中回滚事务，取消所有操作。

假如在`IssueManager.CreateAsync`中先保存一次数据，那么数据会先执行一次*Insert*操作，后面关联用户的逻辑执行后，又会再执行一次*Update*操作。

如果不在`IssueManager.CreateAsync`中保存数据，那么，新建`Issue`和关联用户，只会执行一次*Insert*操作。

##### 讨论：为什么没有在应用服务中执行标题是否重复的检查?

简单地说“因为它是**核心领域逻辑**，应该在领域层实现”。这又带来一个新问题，“**如何确定**是领域层逻辑，还是应用层逻辑”？（这个我们后面再详细讨论）

> 对于此示例，可以用一个简单的问题来判断到底是领域逻辑还是应用逻辑：“如果又另外一种创建`Issue`的方式（用例），我们是否还需要执行？如果需要执行，就属于领域层逻辑，不需要执行就是应用层逻辑”。您可能认为为什么还有别的用例来创建`Issue`呢？

Reviewer: ```suggestion
对于此示例，可以用一个简单的问题来判断到底是领域逻辑还是应用逻辑：“如果还有另外一种创建`Issue`的方式（用例），我们是否还需要执行？如果需要执行，就属于领域层逻辑，不需要执行就是应用层逻辑”。您可能认为为什么还有别的用例来创建`Issue`呢？
```

    private readonly IRepository<AppUser, Guid> _userRepository;

    public IssueAppService(
        IssueManager issueManager,
        IRepository<Issue, Guid> issueRepository,
        IRepository<AppUser, Guid> userRepository)
    {
        _issueManager = issueManager;
        _issueRepository = issueRepository;
        _userRepository = userRepository;
    }

    public async Task<IssueDto> UpdateAsync(Guid id, UpdateIssueDto input)
    {
        // Get entity from database
        var issue = await _issueRepository.GetAsync(id);

        // Change Title
        await _issueManager.ChangeTitleAsync(issue, input.Title);

        // Change Assigned User
        if (input.AssignedUserId.HasValue)
        {
            var user = await _userRepository.GetAsync(input.AssignedUserId.Value);
            await _issueManager.AssignToAsync(issue, user);
        }

        // Change Text (no business rule, all values accepted)
        issue.Text = input.Text;

        // Update entity in the database
        await _issueRepository.UpdateAsync(issue);

        // Return a DTO represents the new Issue
        return ObjectMapper.Map<Issue, IssueDto>(issue);
    }
}
````

> * `UpdateAsync` 方法参数 `id`被作为独立参数，放置在`UpdateIssueDto`之外，当您将此应用服务[自动导出](API/Auto-API-Controllers.md)为HTTP API时，可以帮忙ABP准确判断路由，这与DDD无关。

Reviewer: ```suggestion
* `UpdateAsync` 方法参数 `id`被作为独立参数，放置在`UpdateIssueDto`之外。这是一项设计决策，可在您将服务[自动导出](API/Auto-API-Controllers.md)为HTTP API端点时帮助ABP正确定义HTTP路由。因此这与DDD无关。
```

* **错误**：请勿从应用服务中返回实体，应该返回DTO。

**讨论：为什么不将支付逻辑移到领域服务中？**

您可能想知道为什么付款逻辑代码不在`OrganizationManager`中。付款是非常**重要的事情**，我们不能**遗漏任何一次付款**。

它确实非常重要，但是，它不能放到领域服务中。我们可能还又**其它用例**来创建组织但不收取任何费用。例如：

* 管理员可以在后台管理系统创建新组织，而无需支付任何费用。
* 后台作业系统导入、集成、同步组织而无需支付费用。

如您所见，**付款不是创建有效组织的必要操作**。它是特定的应用服务逻辑。

**示例：CRUD操作**

````csharp
public class IssueAppService
{
    private readonly IssueManager _issueManager;

    public IssueAppService(IssueManager issueManager)
    {
        _issueManager = issueManager;
    }

    public async Task<IssueDto> GetAsync(Guid id)
    {
        return await _issueManager.GetAsync(id);
    }

    public async Task CreateAsync(IssueCreationDto input)
    {
        await _issueManager.CreateAsync(input);
    }

    public async Task UpdateAsync(UpdateIssueDto input)
    {
        await _issueManager.UpdateAsync(input);
    }

    public async Task DeleteAsync(Guid id)
    {
        await _issueManager.DeleteAsync(id);
    }
}
````

该应用服务本身**不执行任何操作**，并将所有**操作转发给** *领域服务*。它甚至将DTO传递给`IssueManager`

* 如果没有**任何业务逻辑**，只有简单的**CRUD**操作，**请勿**创建领域服务。
* **切勿**将**DTOs**传递给领域服务，或从领域服务返回**DTOs**。

可以在应用服务中直接注入仓储，实现查询、创建、更新及删除操作。除非在这些操作过程中需要执行某些业务逻辑，在这种情况下，请创建领域服务。

> > 不要创建这种CRUD领域服务([YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it))，等将来确实需要时，我们再进行重构，重构并不会影响UI或其它客户端。

Reviewer: ```suggestion
> 不要创建“将来可能需要”这种CRUD域服务方法([YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it))！ 在需要时重构它并重构现有代码。 由于应用层优雅地抽象了领域层，因此重构过程不会影响UI层和其他客户端。
```
